### PR TITLE
#338 [Copy] Copying an object with telemetry from the data dictionary results in a number of unknown objects in copied tree

### PR DESCRIPTION
**Pull request by akhenry**
Summary of changes:
* Created a new creation policy category and instance that returns true if tested object hasFeature('creation')`
* Modified the `CopyService` to use the persistence space of the copy target, not the object being copied
* Modified the CreateActionProvider to use new CreationPolicy
* Modified the `CopyService` to use the CreationPolicy to check whether a new object or a link should be created for the type of the object being copied.
* Added tests for CreationPolicy
* Added a new test case to test that duplication of non-createable types creates links (addresses https://github.com/nasa/openmctweb/issues/338#issuecomment-159353004 )
* Some minor refactoring of CopyService and CopyTask. Would suggest that use of instantiationCapability actually allows for more fundamental re-architecture of CopyTask that could significantly reduce LOC, but I'll leave this for a future release.

### Author Checklist

1. Changes address original issue? __Y__
2. Unit tests included and/or updated with changes? __Y__
3. Command line build passes? __Y__
4. Expect to pass code review? __Y__

### CODE REVIEW

File: platform/commonUI/browse/src/creation/CreationPolicy.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    [],
+    function () {
+        "use strict";
+
+        /**
+         * A policy for determining whether objects of a certain type can be
+         * created.
+         * @returns {{allow: Function}}
+         * @constructor
+         */
```
Comment (VWoeltjen) [inline platform/commonUI/browse/src/creation/CreationPolicy.js]: JSDoc here looks autogenerated; for consistency with other sources, I'd use:

```
@constructor
@implements {Policy}
@memberof platform/commonUI/browse
```

File: platform/commonUI/browse/src/creation/CreationPolicy.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        function CreationPolicy() {
+            return {
+                /**
+                 * Only allow creation of object types that have the
+                 * Creation capability
+                 */
+                allow: function (type) {
```
Comment (VWoeltjen) [inline platform/commonUI/browse/src/creation/CreationPolicy.js]: I would move this method to `CreationPolicy.prototype` for consistent code style. On the plus side, JSDoc can be omitted (it's covered by the `Policy` interface)

File: platform/entanglement/src/services/CopyTask.js [hunk 3]
```diff
@@ -185,6 +203,10 @@ define(
         CopyTask.prototype.perform = function(){
             this.deferred = this.$q.defer();
 
+            if (!this.parent.hasCapability('composition')){
```
Comment (VWoeltjen) [inline platform/entanglement/src/services/CopyTask.js]: Out of curiosity, are there cases where this situation happens? I would think that `CopyService.validate` would prevent that

Comment (akhenry) [inline platform/entanglement/src/services/CopyTask.js]: Actually that's a good catch. I moved this from elsewhere in the code, without checking whether it was really necessary, but it is essentially duplicating a check in CopyService.validate. Furthermore, it's not even correct, should be using the policy and not the capability. I'll remove it.

File: platform/commonUI/browse/src/creation/CreationPolicy.js (hunk 2 above)
Comment (akhenry) [inline platform/commonUI/browse/src/creation/CreationPolicy.js]: Fair enough. We should probably also update the code in the examples directory for consistency with this particular style rule. For some reason I used example/policy as the exemplar, and I should have used an active policy.

File: platform/entanglement/src/services/CopyTask.js [hunk 4]
```diff
@@ -33,36 +33,48 @@ define(
...
             //Once copied, associate each cloned
             // composee with its parent clone
-            child.model.location = parent.id;
-            parent.model.composition = parent.model.composition || [];
-            return parent.model.composition.push(child.id);
+
+            parent.getModel().composition.push(child.getId());
+
+            //Check if the object being composed is a link
+            if (!child.getCapability("location").isLink()) {
```
Comment (VWoeltjen) [inline platform/entanglement/src/services/CopyTask.js]: It seems like this check is intended to distinguish cloned objects (whose `location` properties will have been wiped in `cloneObjectModel`.) Since the check isn't testing for exactly that, it could be brittle later - external changes to location could break this, or even maintenance of CopyTask.

I think it would be safer to make the determination about whether or not a location will be set within the block where we've consulted with policy service. Second-best place to do this would be somewhere where we have a reference to both the original and the child (e.g. `copyComposees`), where we can compare to see if this has changed. Could pass in a `setLocation` flag or a `originalChild` object if we want to keep model changes in a small number of locations.

File: platform/entanglement/src/services/CopyTask.js [hunk 5]
```diff
@@ -137,23 +138,37 @@ define(
...
+
+            //Check if the type of the object being copied allows for
+            // creation of new instances. If it does not, then a link to the
+            // original will be created instead.
+            if (this.policyService.allow("creation", originalObject.getCapability("type"))){
+                //create a new clone of the original object. Use the
+                // creation capability of the targetParent to create the
+                // new clone. This will ensure that the correct persistence
+                // space is used.
+                clone = this.parent.hasCapability("instantiation") && this.parent.useCapability("instantiation", cloneObjectModel(originalObject.getModel()));
```
Comment (VWoeltjen) [inline platform/entanglement/src/services/CopyTask.js]: It would be safe to omit the `hasCapability` check here - we're going to fail anyway if this doesn't exist so may as well fail fast. `instantiation` is in core so it's safe to assume it will be there, and if that assumption is ever invalidated I think the check would belong in `CopyService#validate`

File: platform/commonUI/browse/src/creation/CreationPolicy.js (hunk 2 above)
Comment (VWoeltjen) [inline platform/commonUI/browse/src/creation/CreationPolicy.js]: > For some reason I used example/policy as the exemplar

What a silly thing to do! 

Added a note to #45 ("add narrative to example code") indicating that we need to catch up with platform code style, too.

File: platform/entanglement/src/services/CopyTask.js (hunk 4 above)
Comment (akhenry) [inline platform/entanglement/src/services/CopyTask.js]: Hrm, the intention is to literally check whether the object is a link, it's not an indirect way of checking anything else, although you are correct that it will in effect distinguish between cloned and linked objects. Although most links get duplicated as fully-fledged objects (as per https://github.com/nasa/openmctweb/pull/249#discussion_r44058570 ), non-creatable types get composed as links, and we don't want to overwrite their location. If in the future we decided that links should be copied as links as well, this will still work in that instance. In any case, if an object is a link, we don't want to set its location.

I don't think it's particularly brittle, and it made sense to me to do the steps involved in composition in one place. Perhaps its a conceptual misunderstanding though, I tend to think of location being a step in the act of composition (of a new object anyway), but I suppose it's not really?

File: platform/entanglement/src/services/CopyTask.js (hunk 5 above)
Comment (akhenry) [inline platform/entanglement/src/services/CopyTask.js]: Yep, fair enough.

**Comment by akhenry** [comment]
As a general note, the CopyService does not make use of the MutationCapability (either directly, or indirectly via CompositionCapability.add) because this triggers a listener in the `GenericSearchProvider` which then tries to go and fetch the object from its persistence store. Because the objects have not been persisted yet, this results in a lot of 404s in the console. For now I do mutation and composition manually on the models of the copied objects. 

In future it might be a good idea to refactor this to make use of the composition and mutation capabilities.  This would necessitate a change to the `GenericSearchProvider` however. Some investigation should be done into whether the search indexer really needs to fetch the object from persistence again when it has received it via the listener callback anyway. Mutation doesn't even do persistence automatically, so the model it will fetch from persistence could be stale?

**Comment by akhenry** [comment]
Have fixed all of the issues here except https://github.com/nasa/openmctweb/pull/371#discussion_r46633765

File: platform/entanglement/src/services/CopyTask.js (hunk 4 above)
Comment (VWoeltjen) [inline platform/entanglement/src/services/CopyTask.js]: Thanks for the explanation, I understand the rationale for this a little better now. There are two things that still concern me (both of which I think would be easy to address):

1. For cloned objects, we're already making fairly low-level changes to various model properties, so I think it's risky to rely on the capabilities which use those properties to behave correctly. In this case, we've set `location` to undefined, and since there is no `context` for the cloned object, the LocationCapability happens to treat this as a non-link (where is it located? `undefined`. what is its current context? `undefined`.) I think the truth here is that the data model is just incomplete (by necessity - we're in the process of making it complete.) 
2. Since we are making so many low-level changes, I think it's important that we structure the copy procedure such that original objects (not the clones we have created) are well-insulated from those changes. Since `child` in this function may be a non-clone, having a modification without any explicit check to make sure it's a clone that we own seems inherently risky. This check does accomplish that, but that's neither the intent nor the actual logic; we end up relying on the fact that non-cloned objects will always be links, which I believe is true, but there is a chain of assumptions and preconditions that is necessary for that to be true.

In both cases I think this can be addressed trivially:

1. For clones, it should be sufficient to change the `isLink` check to `model.location === undefined`. This matches up nicely with the `delete model.location` that happened at clone-time.
2. For non-clones, passing in a flag to `composeChild` would be fine; we can compare the object we got back from `copy` with the object we put into `copy` and determine if it's a clone at the time `composeChild` gets called.

Comment (akhenry) [inline platform/entanglement/src/services/CopyTask.js]: Ok, fair enough. Incidentally, I've added a comment on the low-level model manipulation and the reason for it. 

Is it even necessary to have the flag? If it's a clone, the location will have been deleted, if it's a link the location will be set to the object's original location. So if it's made it to that function with a location, don't overwrite it?

Comment (akhenry) [inline platform/entanglement/src/services/CopyTask.js]: Tests pass, and from my own testing just checking model.location===undefined is sufficient. Are there edge cases I haven't considered? I am assuming that at the top level all objects are located inside ROOT? Or is this not a safe assumption?

Comment (akhenry) [inline platform/entanglement/src/services/CopyTask.js]: Ignore previous, in the interests of getting this merged before code freeze I've added the setLocation flag.

**Comment by VWoeltjen** [comment]
> Should I ticket this, or am I missing something here?

Yes, I think this approach makes sense. Agreed that there's no reason to go back to `modelService` from the shared worker - the topic should broadcast the changed model (or changed domain object) and generic search should be able to index the new version immediately.

**Comment by VWoeltjen** [comment]
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
